src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,805 @@
+                // items based on a number that exists in the name of the item. However, when
+                // we are not after a dot (i.e. we're being brought up after <space> is typed)
+                // then we don't want to filter things. Consider the user writing:
+                //
+                //      dim i =<space>
+                //
+                // We'll bring up the completion list here (as VB has completion on <space>).
+                // If the user then types '3', we don't want to match against Int32.
+                if (_filterText.Length > 0 && char.IsNumber(_filterText[0]) && !IsAfterDot(_data.Snapshot, _session.ApplicableToSpan))
+                {
+                    // Dismiss the session.
+                    return true;
+                }
+
+                // DismissIfLastCharacterDeleted should be applied only when started with Insertion, and then Deleted all characters typed.
+                // This conforms with the original VS 2010 behavior.
+                if (InitialTriggerReason == CompletionTriggerReason.Insertion &&
+                    UpdateTriggerReason == CompletionTriggerReason.Backspace &&
+                    _completionRules.DismissIfLastCharacterDeleted &&
+                    _filterText.Length == 0)
+                {
+                    // Dismiss the session
+                    return true;
+                }
+
+                return false;
+
+                static bool IsAfterDot(ITextSnapshot snapshot, ITrackingSpan applicableToSpan)
+                {
+                    var position = applicableToSpan.GetStartPoint(snapshot).Position;
+                    return position > 0 && snapshot[position - 1] == '.';
+                }
+            }
+
+            private void AddCompletionItems(List<MatchResult<VSCompletionItem>> list, CancellationToken cancellationToken)
+            {
+                // FilterStateHelper is used to decide wheter a given item should be included in the list based on the state of filter/expander buttons.
+                var filterHelper = new FilterStateHelper(_data.SelectedFilters);
+                filterHelper.LogTargetTypeFilterTelemetry(_session);
+
+                // We want to sort the items by pattern matching results while preserving the original alphabetical order for items with
+                // same pattern match score, but `List<T>.Sort` isn't stable. Therefore we have to add a monotonically increasing integer
+                // to `MatchResult` to keep track the original alphabetical order of each item.
+                var currentIndex = 0;
+
+                // Convert intial and update trigger reasons to corresponding Roslyn type so 
+                // we can interact with Roslyn's completion system
+                var roslynInitialTriggerKind = Helpers.GetRoslynTriggerKind(InitialTriggerReason);
+                var roslynFilterReason = Helpers.GetFilterReason(UpdateTriggerReason);
```

Comment:
i feel like we should do this up front, and only expose the roslyn side of htings.  is there a reason we ever want to work with the non-roslyn side?